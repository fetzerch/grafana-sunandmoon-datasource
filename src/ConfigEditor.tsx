import React, { ChangeEvent, PureComponent } from 'react';
import { InlineField, Input } from '@grafana/ui';
import { DataSourcePluginOptionsEditorProps } from '@grafana/data';
import { SunAndMoonDataSourceOptions } from './types';
import {} from '@emotion/core';

export interface Props extends DataSourcePluginOptionsEditorProps<SunAndMoonDataSourceOptions> {}

export class ConfigEditor extends PureComponent<Props> {
  constructor(props: Props) {
    super(props);

    this.fillPositionFromGeoLocation();
  }

  fillPositionFromGeoLocation = () => {
    const { options } = this.props;
    const { jsonData } = options;
    if (jsonData.latitude === undefined && jsonData.longitude === undefined) {
      if (navigator.geolocation) {
        navigator.geolocation.getCurrentPosition(
          position => {
            const { onOptionsChange, options } = this.props;
            const jsonData = {
              ...options.jsonData,
              latitude: position.coords.latitude,
              longitude: position.coords.longitude,
            };
            onOptionsChange({ ...options, jsonData });
          },
          error => {},
          {}
        );
      }
    }
  };

  onLatitudeChange = (event: ChangeEvent<HTMLInputElement>) => {
    const value = parseFloat(event.target.value);
    const { onOptionsChange, options } = this.props;
    const jsonData = {
      ...options.jsonData,
      latitude: isNaN(value) ? undefined : value,
    };
    onOptionsChange({ ...options, jsonData });
  };

  onLongitudeChange = (event: ChangeEvent<HTMLInputElement>) => {
    const value = parseFloat(event.target.value);
    const { onOptionsChange, options } = this.props;
    const jsonData = {
      ...options.jsonData,
      longitude: isNaN(value) ? undefined : value,
    };
    onOptionsChange({ ...options, jsonData });
  };

  render() {
    const { options } = this.props;
    const { jsonData } = options;

    return (
      <div className="gf-form-group">
        <h3 className="page-heading">Sun and Moon reference location</h3>
        <div className="gf-form">
          <InlineField label="Latitude" labelWidth={14}>
            <Input
              className="latitude"
              label="Latitude"
              onChange={this.onLatitudeChange}
              value={jsonData.latitude}
              placeholder="latitude"
              type="number"
              min={-90}
              max={90}
              required
            />
          </InlineField>
        </div>

        <div className="gf-form">
          <InlineField label="Longitude" labelWidth={14}>
            <Input
              className="longitude"
              label="Longitude"
              onChange={this.onLongitudeChange}
              value={jsonData.longitude}
              placeholder="longitude"
              type="number"
              min={-360}
              max={360}
              required
            />
          </InlineField>
        </div>
      </div>
    );
  }
}
