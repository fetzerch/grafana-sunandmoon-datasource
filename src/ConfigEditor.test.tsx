import React from 'react';
import { shallow } from 'enzyme';
import { ConfigEditor } from './ConfigEditor';

const setup = (propsOverrides?: object) => {
  const props: any = {
    options: {
      jsonData: {
        latitude: 48,
        longitude: 10,
      },
    },
    onOptionsChange: jest.fn(),
  };

  Object.assign(props, propsOverrides);
  return shallow(<ConfigEditor {...props} />);
};

describe('Render', () => {
  it('should render component', () => {
    const wrapper = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('should initially set latitude and longitude from geolocation', () => {
    const onOptionsChange = jest.fn();
    setup({ onOptionsChange: onOptionsChange, options: { jsonData: {} } });
    expect(onOptionsChange).not.toHaveBeenCalled();

    const mockGeolocation = {
      getCurrentPosition: jest.fn().mockImplementationOnce(success =>
        Promise.resolve(
          success({
            coords: {
              latitude: 40,
              longitude: 5,
            },
          })
        )
      ),
      clearWatch: jest.fn(),
      watchPosition: jest.fn(),
    };
    Object.defineProperty(global.navigator, 'geolocation', { value: mockGeolocation });
    setup({ onOptionsChange: onOptionsChange, options: { jsonData: {} } });
    expect(onOptionsChange).toHaveBeenLastCalledWith({ jsonData: { latitude: 40, longitude: 5 } });
  });

  it('should update latitude and longitude on change', () => {
    const onOptionsChange = jest.fn();
    const wrapper = setup({ onOptionsChange: onOptionsChange });
    const latitudeField = wrapper.find('.latitude');
    const longitudeField = wrapper.find('.longitude');

    latitudeField.simulate('change', { target: { value: 50 } });
    expect(onOptionsChange).toHaveBeenLastCalledWith({ jsonData: { latitude: 50, longitude: 10 } });
    longitudeField.simulate('change', { target: { value: 20 } });
    expect(onOptionsChange).toHaveBeenLastCalledWith({ jsonData: { latitude: 48, longitude: 20 } });
    latitudeField.simulate('change', { target: { value: '' } });
    expect(onOptionsChange).toHaveBeenLastCalledWith({ jsonData: { latitude: undefined, longitude: 10 } });
    longitudeField.simulate('change', { target: { value: '' } });
    expect(onOptionsChange).toHaveBeenLastCalledWith({ jsonData: { latitude: 48, longitude: undefined } });
  });
});
