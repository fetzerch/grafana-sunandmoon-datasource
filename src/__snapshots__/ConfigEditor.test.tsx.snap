// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render should render component 1`] = `
<div
  className="gf-form-group"
>
  <h3
    className="page-heading"
  >
    Sun and Moon reference location
  </h3>
  <div
    className="gf-form"
  >
    <InlineField
      label="Latitude"
      labelWidth={14}
    >
      <Input
        className="latitude"
        label="Latitude"
        max={90}
        min={-90}
        onChange={[Function]}
        placeholder="latitude"
        required={true}
        type="number"
        value={48}
      />
    </InlineField>
  </div>
  <div
    className="gf-form"
  >
    <InlineField
      label="Longitude"
      labelWidth={14}
    >
      <Input
        className="longitude"
        label="Longitude"
        max={360}
        min={-360}
        onChange={[Function]}
        placeholder="longitude"
        required={true}
        type="number"
        value={10}
      />
    </InlineField>
  </div>
</div>
`;
